libbeat/outputs/mode/balance_async.go

```diff
@@ -0,0 +1,307 @@
+	// workers become unresponsive.
+	work    chan eventsMessage
+	retries chan eventsMessage
+}
+
+// NewAsyncLoadBalancerMode create a new load balancer connection mode.
+func NewAsyncLoadBalancerMode(
+	clients []AsyncProtocolClient,
+	maxAttempts int,
+	waitRetry, timeout, maxWaitRetry time.Duration,
+) (*AsyncLoadBalancerMode, error) {
+
+	// maxAttempts signals infinite retry. Convert to -1, so attempts left and
+	// and infinite retry can be more easily distinguished by load balancer
+	if maxAttempts == 0 {
+		maxAttempts = -1
+	}
+
+	m := &AsyncLoadBalancerMode{
+		timeout:      timeout,
+		maxWaitRetry: maxWaitRetry,
+		waitRetry:    waitRetry,
+		maxAttempts:  maxAttempts,
+
+		work:    make(chan eventsMessage),
+		retries: make(chan eventsMessage, len(clients)*2),
+		done:    make(chan struct{}),
+	}
+	m.start(clients)
+
+	return m, nil
+}
+
+// Close stops all workers and closes all open connections. In flight events
+// are signaled as failed.
+func (m *AsyncLoadBalancerMode) Close() error {
+	close(m.done)
+	m.wg.Wait()
+	return nil
+}
+
+// PublishEvents forwards events to some load balancing worker.
+func (m *AsyncLoadBalancerMode) PublishEvents(
+	signaler outputs.Signaler,
+	opts outputs.Options,
+	events []common.MapStr,
+) error {
+	return m.publishEventsMessage(opts,
+		eventsMessage{signaler: signaler, events: events})
+}
+
+// PublishEvent forwards the event to some load balancing worker.
+func (m *AsyncLoadBalancerMode) PublishEvent(
+	signaler outputs.Signaler,
+	opts outputs.Options,
+	event common.MapStr,
+) error {
+	return m.publishEventsMessage(opts,
+		eventsMessage{signaler: signaler, event: event})
+}
+
+func (m *AsyncLoadBalancerMode) publishEventsMessage(
+	opts outputs.Options,
+	msg eventsMessage,
+) error {
+	maxAttempts := m.maxAttempts
+	if opts.Guaranteed {
+		maxAttempts = -1
+	}
+	msg.attemptsLeft = maxAttempts
+
+	if ok := m.forwardEvent(m.work, msg); !ok {
+		dropping(msg)
+	}
+	return nil
+}
+
+func (m *AsyncLoadBalancerMode) start(clients []AsyncProtocolClient) {
+	var waitStart sync.WaitGroup
+	worker := func(client AsyncProtocolClient) {
+		defer func() {
+			if client.IsConnected() {
+				_ = client.Close()
+			}
+			m.wg.Done()
+		}()
+
+		waitStart.Done()
+
+		backoff := newBackoff(m.done, m.waitRetry, m.maxWaitRetry)
+		for {
+			// reconnect loop
+			for !client.IsConnected() {
+				if err := client.Connect(m.timeout); err == nil {
+					break
+				}
+
+				if !backoff.Wait() { // done channel closed
+					return
+				}
+			}
+
+			// receive and process messages
+			var msg eventsMessage
+			select {
+			case <-m.done:
+				return
+			case msg = <-m.retries: // receive message from other failed worker
+				debug("events from retries queue")
+			case msg = <-m.work: // receive message from publisher
+				debug("events from worker worker queue")
+			}
+
+			err := m.onMessage(client, msg)
+			if !backoff.WithError(err) { // done channel closed
```

Comment:
The comment should probably read "done, channel closed" for clarity.

I know it's not part of this PR, but I think the `WithError` method would have been better called `WaitUnlessError` or something like that.